Fix modbus climate precision configuration variable

Kirchoff opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
This pull request is to better handle the `precision` configuration variable inside the modbus climate module. This module used to do rounding itself before storing to either the current or target temperature, instead it should use the built-in `climate` precision property. Because this module did not use the precision property, home assistant was assuming that the precision was WHOLE numbers, regardless of what the configuration variable was set to. It is for that reason I am marking this as a bugfix PR.
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Kirchoff committed:
Updated precision to follow the tenths, halves, whole notation used by
other home assistant climate modules. Added the precision @property so
that home assistant can handle this rounding in the frontend, rather
than in the _async_read_register() method.

Kirchoff committed:
Merge branch 'home-assistant:dev' into dev_modbus_climate_bugfix

Kirchoff committed:
Fixed a pylinter error for periods in user-facing log messages, and
updated `precision` defaults in components/modbus/__init__.py to be
consistent with an error case, using `PRECISION_WHOLE`.

janiversen commented:
You have a point, but you forget the "precision:" is common between sensor and climate, so when you change the schema you force sensor to use the same pattern.

We use precision to set the number of decimals wanted, and as far as I can see (after a quick glance) your solution does not provide that possibility.

If "precision:" causes problems in the UI, then of course a change is needed, but it needs to allow us to configure number of digits as well.

Kirchoff commented:
@janiversen - Thank you for your review and consideration. If you have suggestions for how we can preserve functionality of both, I'm open to suggestions. I assumed adding an additional new configuration variable was possible but not preferred. 

I purposely did not change the schema validation inside __init__.py, `BASE_STRUCT_SCHEMA`, and instead added it as an optional configuration variable for climate under `CLIMATE_SCHEMA`. It's my understanding that this will not affect the functionality of sensors (still `cv.positive_int`), only "override" it for climate (`vol.Coerce(float)`).

I will add some sensors to my development configuration and see if I'm able to test whether or not this is the case.

janiversen commented:
I did not do a review I postponed that until the principal problems are solved.

Having “precision:” in the climate schema overwriting the struct schema is confusing and will not pass review. I am sure with a proper schema it is possible to have both, but that would also allow float values for climate, which I think is correct to have.

Kirchoff commented:
I will return the `precision:` variable to it's intended use. We should be able to infer precision for the climate `precision()` method based on whether our modbus `precision:` is greater than or equal to 1. In this case we can support `PRECISION_TENTHS` and `PRECISION_WHOLE`.

Is that acceptable?

janiversen commented:
That sounds acceptable.

Kirchoff committed:
revert changes to `precision:` configuration variable
instead, the climate `precision()` function will infer whether or not
to display in whole or tenths. halves will be unsupported, which should
be fine.

Kirchoff committed:
re-added missing line that was removed

Kirchoff committed:
revert change to use self._input_type instead of CALL_TYPE_REGISTER_HOLDING

Kirchoff commented:
Should be fixed.

janiversen reviewed (approved):
Look good to me.

janiversen commented:
You need to resolve the change requests.
